module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,103 @@
+#include <cmath>
+#include <iterator>
+#include <sstream>
+
+#include "advancedCalculator.hpp"
+
+const std::map<char, std::function<double(double, double)>> calculations {
+    {'+', std::plus<>()},
+    {'-', std::minus<>()},
+    {'*', std::multiplies<>()},
+    {'/', std::divides<>()},
+    {'%', std::modulus<int>()},
+    {'!', [](auto n, auto x) { return n <= 0 ? 1.0 : std::tgamma(n + 1.0 + x - x); }},
+    {'^', [](auto base, auto exponent) { return pow(base, exponent); }},
+    {'$', [](auto value, auto root) { return pow(value, 1.0 / root); }}
+};
+
+bool isFormatBad(const std::vector<std::string>& terms) {
+    return ((terms.size() < MAX_TERMS) ||
+            (terms.size() > MAX_TERMS + 1) ||
+            (terms.at(OPERATOR).length() != 1) ||
+            (!std::regex_search(terms.at(OPERATOR), arithmeticOp) &&
+                !std::regex_search(terms.at(OPERATOR), factorialOp)) ||
+            (!(terms.size() <= MAX_TERMS) &&
+                (std::regex_search(terms.at(OPERATOR), factorialOp))) ||
+            (terms.front().length() == OPERATOR &&
+                std::regex_search(terms.front(), arithmeticOp)) ||
+            (terms.back().length() == OPERATOR &&
+                std::regex_search(terms.back(), arithmeticOp))
+       ) ? true : false;
+}
```

Comment:
It looks a bit complicated at least for me, but I don't have any idea how to make it simpler

Reply:
Yes, it is complicated - the result of TDD. I had to add more and more if statements to pass the tests. Refactoring could help but not that knowledge level yet. Thanks!